src/Security/Authentication/Negotiate/src/Internal/LdapAdapter.cs

```diff
@@ -45,24 +64,35 @@ public static async Task RetrieveClaimsAsync(LdapSettings settings, ClaimsIdenti
 
                     if (!settings.IgnoreNestedGroups)
                     {
-                        GetNestedGroups(settings.LdapConnection, identity, distinguishedName, groupCN, logger);
+                        GetNestedGroups(settings.LdapConnection, identity, distinguishedName, groupCN, logger, retrievedClaims);
                     }
                     else
                     {
-                        AddRole(identity, groupCN);
+                        retrievedClaims.Add(CreateClaim(identity, groupCN));
                     }
                 }
+
+                foreach (var claim in retrievedClaims)
+                {
+                    identity.AddClaim(claim);
+                }
+
+                settings.ClaimsCache.Set(user,
+                    retrievedClaims,
+                    new MemoryCacheEntryOptions()
+                        .SetSize(1)
```

Comment:
This can be measured a lot more accurately, the claims are just strings. Users are also going to have very different size lists and should be weighted accordingly.

Try totaling the lengths of the strings involved in the above foreach enumeration.